### PR TITLE
Data Transfer Statistics: number of transfers

### PR DESCRIPTION
**Pull request by srdieffenbach**
Number of transfers tracked in Data Transfer server. 
@nichhk

### CODE REVIEW

File: rice-datanode/source/data_transfer_server.cc [hunk 1]
```diff
@@ -70,11 +74,13 @@ void TransferServer::handle_connection(tcp::socket sock) {
 }
 
 void TransferServer::processWriteRequest(tcp::socket& sock) {
+	xmits++;
```
Comment (pelmers) [inline rice-datanode/source/data_transfer_server.cc]: Can the xmit operations all move to handle_connection?
Increment on calling write/read, decrement on return.

File: rice-datanode/include/data_transfer_server.h [hunk 2]
```diff
@@ -38,11 +38,12 @@ class TransferServer {
 	public:
 		TransferServer(int port, nativefs::NativeFS& fs, zkclient::ZkClientDn& dn);
 		void serve(asio::io_service& io_service);
+		int getNumTransmits(void);
```
Comment (nichhk) [inline rice-datanode/include/data_transfer_server.h]: This should be deleted.

File: rice-datanode/source/data_transfer_server.cc [hunk 3]
```diff
@@ -24,6 +24,7 @@ bool TransferServer::receive_header(tcp::socket& sock, uint16_t* version, unsign
 }
 
 void TransferServer::handle_connection(tcp::socket sock) {
+    using namespace std::placeholders;
```
Comment (nichhk) [inline rice-datanode/source/data_transfer_server.cc]: This should be deleted.

File: rice-datanode/source/data_transfer_server.cc [hunk 4]
```diff
@@ -36,12 +36,16 @@ void TransferServer::handle_connection(tcp::socket sock) {
...
-				processWriteRequest(sock);
-				break;
-			case (READ_BLOCK):
-				processReadRequest(sock);
-				break;
+			case (WRITE_BLOCK): {
+                std::function <void(TransferServer&, tcp::socket&)> writeFn = &TransferServer::processWriteRequest;
+                synchronize(writeFn, sock);
+            }
+            break;
```
Comment (pelmers) [inline rice-datanode/source/data_transfer_server.cc]: should the break be outside the braces?

File: rice-datanode/include/data_transfer_server.h [hunk 5]
```diff
@@ -54,7 +54,8 @@ class TransferServer {
 		bool writeFinalPacket(tcp::socket& sock, uint64_t, uint64_t);
 		template <typename BufType>
 		bool writePacket(tcp::socket& sock, PacketHeaderProto p_head, const BufType& payload);
-};
+        void synchronize(std::function<void(TransferServer&, tcp::socket&)> f, tcp::socket& sock);
```
Comment (pelmers) [inline rice-datanode/include/data_transfer_server.h]: spaces

File: rice-datanode/include/data_transfer_server.h [hunk 6]
```diff
@@ -54,7 +54,8 @@ class TransferServer {
...
+    };
```
Comment (pelmers) [inline rice-datanode/include/data_transfer_server.h]: dedent

File: rice-datanode/source/data_transfer_server.cc [hunk 7]
```diff
@@ -36,12 +36,16 @@ void TransferServer::handle_connection(tcp::socket sock) {
 		}
 		// TODO: implement proto handlers based on type
 		switch (type) {
-			case (WRITE_BLOCK):
-				processWriteRequest(sock);
-				break;
-			case (READ_BLOCK):
-				processReadRequest(sock);
-				break;
+			case (WRITE_BLOCK): {
```
Comment (pelmers) [inline rice-datanode/source/data_transfer_server.cc:39]: spaces

File: rice-datanode/source/data_transfer_server.cc [hunk 8]
```diff
@@ -288,3 +294,9 @@ void TransferServer::serve(asio::io_service& io_service) {
 		std::thread(&TransferServer::handle_connection, this, std::move(sock)).detach();
 	}
 }
+
+void TransferServer::synchronize(std::function<void(TransferServer&, tcp::socket&)> f, tcp::socket& sock){
```
Comment (pelmers) [inline rice-datanode/source/data_transfer_server.cc]: spaces

**Review by pelmers — COMMENTED** [review]
whitespaces

File: rice-datanode/source/data_transfer_server.cc (hunk 4 above)
Comment (nichhk) [reply rice-datanode/source/data_transfer_server.cc]: I don't think it makes a difference. The braces are just to confine the *Fn variables.

**Pull request merged**